README
[ COPYRIGHT NOTICES ]
    OpenGL Image library v1.0
    
    GLi and the GLi specification are (C) Frank L. Morales II (altf2o)
    Source code for libGLi (C) Frank L. Morales II (altf2o)
    
    TGA, TARGA, and the TGA specification are (C) Truevision.
    Source code for libTGA (C) Frank L. Morales II (altf2o)
    
    This code makes use of the zlib lossless compression library.
    Information on the zlib library and it's author's can be 
    found at: http://www.gzip.org/zlib/
    
    The OpenGL name and API is copyright it's respective owners.
    Please visit:  http://www.opengl.org/ for more information.
    
    This library targets the OpenGL API but is in no way associated with
    OpenGL or the OpenGL ARB (Architecture Review Board). This is a 100% 3rd
    party and independantly developed image file format and library. I make
    no claims that myself, the name GLi, the GLi specification and libGLi are 
    in any way associated with OpenGL or the OpenGL ARB.

    
[ GOALS OF THIS LIBRARY ]

    The goal was to create an image format that could easily be used in 
    OpenGL texture mapping. I wanted a library that could compress using
    lossless compression techniques (by the well established zlib library)
    as well as have a VERY easy to understand and parse structure. All 
    while keeping the sections that were important to a modern day image
    file format. We also needed a file format where no section was optional.
    This would reduce the robustness of the file format in comparison to 
    some others. However it would greatly reduce the complexity of parsing the 
    image, which would lead to improved performance and smaller library code.
    The result is the GLi specification. The goals and how i feel i 
    accomplished them are as follows:
    
    1) We needed to be able to quickly and easily determine if the current
    file is a valid GLi image. Therefor i took a note from the 'ELF' 
    executable file format and used a similar 4 byte field. The first 
    variable in our image header is: 
        mBytes (4 bytes). 
    Our images (if they're valid) start w/ the first 4 bytes being:  
        0x7F, 'G', 'L', 'i' . 
    This allows us to quickly determine if the file is valid or not and
    act accordingly. To make it easier we declare an array (in GLi.h)
    which contains these 4 bytes called, MAGIC_BYTES. Using any type of
    memory compare function we can quickly and easily determine if the
    current file is valid.

    2) We needed a way to quickly and most importantly RELIABLY establish the
    EXACT version of the image format used to write the image we're trying to
    work with. We introduce the next variable in our header: 
        version (1 byte). 
    This will contain the version used to write the GLi image. To make it
    easier, this value can be set with, and compared against the #define
    from GLi.h:
        GLI_CURRENT_VERSION
    As of this writing the ONLY possible value is 1.

    3) Next we needed a way to specify a meaningful value to tell OpenGL which
    format our pixel data is in. We introduce the next variable in our header:
        format (4 bytes). 
    To keep things simple we addressed the problem of reversed pixel data, i.e.
    BGR form as apposed to the standard RGB. To make it easier our format ONLY
    supports storing pixel data in the RGB format. As of version 1.0 we support
    8\24\32 bits per pixel. We offer 2 macros in GLi.h the first being:
        CONVERTBGRtoRGB 
    which expects a pointer to a valid 'gli_image' structure. And:
        CONVERTBGRAtoRGBA
    which also expects a pointer to a valid 'gli_image' structure. Both of which
    simply reverse the image data's pixel order.

    4) The next 3 fields are the images "width":
        width (4 bytes)
    "height":
        height (4 bytes) 
    and "pixel depth":
        pxlDepth (1 bytes).
    Those are obviously a must for many things. One issue i had was other 
    formats storing the pixel depth in 'bits per pixel' i.e. for a 24bpp image 
    pxlDepth would be 24. In the GLi format we store the pixel depth in bytes. 
    So for a 24bpp image, pxlDepth will be 3 (8bits per pixel).

    5) I wanted a way to quickly get the size of the raw uncompressed image 
    data. width * height * pxlDepth make up this value. However instead of 
    having to compute that on load, we have a variable: 
        imgSize (4 bytes)
    which holds the size of our raw uncompressed pixel data in bytes.

    6) The next 2 fields weren't really problems in other formats, this is just 
    how i decided they would work best. It's easier IMHO to have 2 valid 
    C-strings of known size to give the author of the image their due credit, 
    as well as a time stamp for when the image was created. This gives us our 
    next two fields:  
        author (24 bytes)
    and: 
        stamp (28 bytes)
    Both can hold 1 less than the given value, that is to save 1 byte for the 
    terminating null character.

    7) We wanted to be able to compress the pixel data as much as possible. 
    However since the target was texture mapping and such in OpenGL the last 
    thing we want is lossy compression. Therefor we acknowledge the fact that 
    we won't get as good of compression ratios, but we can still greatly reduce 
    the size of our images raw pixel data, and keep the same excellent quality. 
    To do that we compress JUST the pixel data with the zlib lossless 
    compression library. (NOTE: Upon loading an image the pixel data is ALWAYS 
    uncompressed. When the function 'gli_load' returns a valid structure the 
    image data has already been uncompressed for you, assuming the image was 
    compressed.) To signal we have compressed data our header structures
    variable: 
        comp (comp[ressed]) (1 byte) 
    will be set to:
        GLI_IMAGE_COMPRESSED. 
    Otherwise it should hold the value: 
        GLI_IMAGE_NOTCOMPRESSED. 
    NOTE: Any other values (besides GLI_IMAGE_COMPRESSED) are assumed to mean 
    the image is NOT compressed. If we do have GLI_IMAGE_COMPRESSED then our 
    next headers variable:  
        compSize (comp[ressed]Size) (4 bytes)
    will hold the size of our compressed pixel data (in bytes). NOTE: We also 
    specify 2 #define's named: 
        GLI_IMAGE_COMPRESS
    and it's counterpart:
        GLI_IMAGE_DONTCOMPRESS
    They should be used when telling a function who's preparing an image 
    structure, to compress the data, or not to compress the data. That function 
    should then set the structures 'comp' variable to: GLI_IMAGE_COMPRESSED, 
    or GLI_IMAGE_NOTCOMPRESSED respectively. 

---------- Up until now this is what we consider our gli_image structures 
           "header" it is a total of 79 bytes. After this there's only 1 more 4 
           byte variable. When reading in the entire header do NOT include that 
           4 byte variable. To make that easier to remember ALWAYS use the 
           supplied #define:
                GLI_HEADER_SIZE .
           
    8) Lastly we wanted our pixel data to ALWAYS be in a known place. We 
    therefor didn't include any optional sections like some other formats do. 
    So the last thing our structure has is a variable named: 
        image (4 bytes). 
    Which is a pointer to some area of memory we'll allocate and store the 
    (raw) pixel data into.
    
    NOTE: This format was an experiment i happened to find useful. It was not
    meant to be a replace all, widely used image format. I could have used a
    stripped targa loader, or bitmap loader etc... but where's the fun in that?
    :-) If it does become popular great, i'd be glad to see other's benefit from
    my work in any way shape or form. However let it be known my intentions 
    were to create a very strait forward, simple, easy to use and understand,
    image file format for my texture mapping needs. I feel this has accomplished
    that goal quite well.
    
You may modify and or use this library in any way you see fit so long 
as the following is met:
        
[ USING AND MODIFYING THE LIBRARY ]
  The libraries presented here (libGLi and libTGA), and all other accompanying
  "tools" (in the tools_src/ directory) are subject to the LGPL license. A copy
  of that has been shipped w/ this library, and is in the file: LGPL.license
  Please refer to it for rules about modifying this code, and using any of these
  libraries in your free, or proprietary applications.
    
Constructive critisism is always welcome. Please send comments and or questions
to:  altf2o@gmail.com , please remember to use: GLi Comment\Question  in the
"Subject" line of the email. Makes my life easier. That aside, i hope someone
finds this useful, i currently use it for all my texture mapping needs.

FMorales...


ChangeLog
2005-03-05  Frank L. Morales II (altf2o)  <altf2o@comcast.net>

* libGLi_src/libGLi.c (gli_load): Changed the logic in a test, to test for the
  most /expected/ condition. It helped make the flow more obvious, and removed
  a conditional call to 'free' at the end of the code.
  
* libGLi_src/libGLi.c (gli_compress): Re-arranged the way the call to the 
  'compress2' function looks.
  
* libGLi_src/libGLi.c (gli_uncompress): Re-arranged the way the call to the
  'uncompress' function looks.
  
* libGLi_src/libGLi.c (gli_writeOGLBuffer): Re-arranged the way i called the
  'gli_compress' function.
  
* ALL *.h,*.c files: Added a "File ID:" (given by subversion) for every source
  and header file, in the comment header of each file.

2005-03-03  Frank L. Morales II (altf2o)  <altf2o@comcast.net>

* tools_src/tools_convertTGAtoGLi.c (main): Fixed a potential bug caused by 
  a call to 'memcpy' where the second parameter /may/ be less than the specified 
  size.
  
* libGLi_src/libGLi.c (gli_writeOGLBuffer): Fixed a potential bug caused by a 
  call to 'memcpy' where the second parameter /may/ be less then the specified 
  size.
  
2005-03-03  Frank L. Morales II (altf2o)  <altf2o@comcast.net>

* ChangeLog (ChangeLog): Started this changelog :) Lets see if i can stay
  somewhat organized.


include/GLi.h
/*
 * libGLi - A library for loading\saving etc... GLi images
 * Copyright (C) 2005  Frank L. Morales II (altf2o)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * File:
 *  libGLi.h
 *
 * Purpose:
 *  This defines the public API for our libGLi library.
 *
 * File ID:
 *  $Id: GLi.h 8 2005-03-05 23:43:06Z altf2o $
 */     

#ifndef GLI_H
#define GLI_H

/* If we're on Winders, we can make this library a DLL */
#if defined ( WIN32 )
# define DLLEXPORT __declspec( dllexport )
# include <gl/glut.h>
# include <gl/glext.h>
#elif defined ( __FreeBSD__ )
# define DLLEXPORT
# include <GL/glut.h>
# include <GL/glext.h>
#else
# define DLLEXPORT /* This is for Linux, what should we test for? */
# include <GL/glut.h>
# include <GL/glext.h>
#endif

#include <zlib.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* This will convert our pixel data from: BGR form, to RGB. */
#define CONVERTBGRtoRGB( img )            \
  unsigned char bgr_byte;                 \
  unsigned int  u;                        \
  for( u = 0; u < img->imgSize; u += 3 )  \
  {                                       \
    bgr_byte = img->image[u];             \
    img->image[u] = img->image[u + 2];    \
    img->image[u + 2] = bgr_byte;         \
  }                                       
      
/* This will convert our pixel data from: BGRA form, to RGBA. */
#define CONVERTBGRAtoRGBA( img )          \
  unsigned char bgr_byte;                 \
  unsigned int  u;                        \
  for( u = 0; u < img->imgSize; u += 4 )  \
  {                                       \
    bgr_byte = img->image[u];             \
    img->image[u] = img->image[u + 2];    \
    img->image[u + 2] = bgr_byte;         \
  }

/* Size of our header MINUS the 4 byte pointer for 'image' */
#define GLI_HEADER_SIZE 79

/* Right now this MUST be 1 */
#define GLI_CURRENT_VERSION 1

/* All of our images are stored using: RGB\RGBA format. 
 * Or single byte per pixel luminace. 
 */
#define GLI_FORMAT_STANDARD GL_RGB
#define GLI_FORMAT_32BIT    GL_RGBA
#define GLI_FORMAT_8BIT     GL_LUMINANCE

/* Just some meaningful return values for our functions to use */
#define GLI_OK    0
#define GLI_ERROR !GLI_OK

/* These let us know if we should compress the image, and if it is 
 * /already/ compressed etc... 
 */
#define GLI_IMAGE_COMPRESS      0
#define GLI_IMAGE_DONTCOMPRESS  1
#define GLI_IMAGE_COMPRESSED    2
#define GLI_IMAGE_NOTCOMPRESSED 3

/* Every valid GLi image MUST contain these as it's first 4 bytes */
static unsigned const char MAGIC_BYTES[] = { 0x7F, 'G', 'L', 'i' };

/* Your compiler must support this directive for this library to properly
 * work. If it does not, please change this #pragma to something your
 * compiler understands so it keeps our structure "tightly" packed.
 */
#pragma pack(1)
typedef struct _gli_image 
{
  unsigned char  mBytes[4];
  unsigned char  version;
  unsigned int   format;
  unsigned int   width;
  unsigned int   height;
  unsigned char  pxlDepth;
  unsigned int   imgSize;
  char           author[24];
  char           stamp[28];
  unsigned char  comp;
  unsigned int   compSize;
  char          *image;
} gli_image;
#pragma pack(4)

/* Use extern "C" here so we can call these from C++ programs */
#ifdef __cplusplus
extern "C" {
#endif

/* Public API function prototypes */
DLLEXPORT gli_image *gli_load          (const char *in_file);
DLLEXPORT int        gli_write         (const gli_image *img, 
                                        const char      *out_file);
DLLEXPORT void       gli_destroy       (gli_image **img);

DLLEXPORT int        gli_compress      (char              *dest_buff,
                                        unsigned long int *dest_len,
                                        const char        *src_buff,
                                        unsigned long int  src_len);

DLLEXPORT int        gli_uncompress    (char              *dest_buff,
                                        unsigned long int *dest_len,
                                        const char        *src_buff,
                                        unsigned long int  src_len);
DLLEXPORT int        gli_writeOGLBuffer(const char    *out_file, 
                                        const char    *author, 
                                        const char    *stamp,
                                        unsigned char  compress);
#ifdef __cplusplus
}
#endif

#endif /* not GLI_H */


include/TGA.h
/*
 * libTGA - Minimalist library to load a select few targa images
 * Copyright (C) 2005  Frank L. Morales II (altf2o)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * File:
 *  libTGA.h
 *
 * Purpose:
 *  This defines the public API for our libTGA library
 *
 * File ID:
 *  $Id: TGA.h 4 2005-03-04 10:36:51Z altf2o $
 */
#ifndef TGA_H
#define TGA_H

#if defined ( WIN32 )
# define DLLEXPORT __declspec( dllexport )
# include <windows.h>
# include <gl/glut.h>
# include <gl/glext.h>
#elif defined ( __FreeBSD__ )
# define DLLEXPORT
# include <GL/glut.h>
# include <GL/glext.h>
# include <sys/types.h>
# include <sys/stat.h>
#endif

#include <GLi.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


/* Keep ALL of these structures tightly packed so we can
 * safely read the entire thing from the file in a SINGLE
 * fread statement.
 */
#pragma pack(1)
typedef struct _tga_image 
{
  /* Header */
  unsigned char   idLength;
  unsigned char   colorMapType;
  unsigned char   imageType;
    
  /* Color map Specification */
  unsigned short  firstEntryIndex;
  unsigned short  colorMapLength;
  unsigned char   cMapEntrySize;
   
  /* Image specification */
  unsigned short  xorigin;
  unsigned short  yorigin;
  unsigned short  width;
  unsigned short  height;
  unsigned char   pxlDepth;
  unsigned char   imgDescriptor;
    
  /* Various helpful variables */
  unsigned int    fileSize;
  unsigned int    cMapSize;
  unsigned int    imgOffset;
  unsigned int    imgSize;
  unsigned char  *image;
    
  /* Footer */
  unsigned int    extOffset;
  unsigned int    devDirOffset;
    
  /* This should hold:  TRUEVISION-XFILE.0  where 0 is a binary zero */
  unsigned char   signature[17];
} tga_image;
#pragma pack(4)
/* End of tightly packed structures */

#define TGA_TYPE_TRUE_COLOR 2

DLLEXPORT tga_image *tga_load( const char *in_file );
DLLEXPORT void       tga_destroy( tga_image **my_tga );

#endif /* not TGA_H */


libTGA_src/libTGA.c
/*
 * libTGA - Minimalist library to load a select few targa images
 * Copyright (C) 2005  Frank L. Morales II (altf2o)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * File:
 *  libTGA.c
 *
 * Purpose:
 *  This implements the public API we use in libTGA.
 *
 * File ID:
 *  $Id: libTGA.c 8 2005-03-05 23:43:06Z altf2o $
 */ 
#include <TGA.h>


/*
 * Arguments:
 *  in_file - A valid C-Strings containing the name of the TGA image to load.
 *
 * Purpose:
 *  This function loads up a TGA(TARGA) image, and stores it's information
 *  into a dynamically allocated structure (img). Which will get returned.
 *
 * Returns:
 *   img - If everything went ok.
 *  NULL - On error.
 */
DLLEXPORT tga_image *
tga_load(const char *in_file)
{
  tga_image *img;
  FILE      *fin;
    
  if(!in_file)
    return NULL;
    
  img = malloc(sizeof(*img));
  if(img == NULL)
    return NULL;
        
  memset(img, 0, sizeof(*img));
      
  fin = fopen(in_file, "rb");  
  if(fin == NULL) 
  {
    free(img);
    return NULL;
  }
    
  if(fread(img, 18, 1, fin) < 1)
    goto error;
        
  /* We only understand True Color non RLE compressed targa images */
  if(img->imageType != TGA_TYPE_TRUE_COLOR)
    goto error;
    
  img->pxlDepth /= 8;
    
  /* We only understand 8\24\32bit images. */
  if(img->pxlDepth != 3 && img->pxlDepth != 4 && img->pxlDepth != 1)
    goto error;
        
  img->imgSize = img->width * img->height * img->pxlDepth;
    
  img->image = malloc(img->imgSize * sizeof(*img->image));
  if(img->image == NULL)
    goto error;
    
  /* A platform specific, but reliable way to get our file size */
#if defined (WIN32)
  HANDLE targa;
    
  targa = CreateFile(in_file, 
                     GENERIC_READ, 
                     FILE_SHARE_READ, 
                     NULL,
                     OPEN_EXISTING, 
                     FILE_ATTRIBUTE_NORMAL, 
                     NULL);              
  if(targa == INVALID_HANDLE_VALUE)
    goto error;
    
  img->fileSize = GetFileSize(targa, NULL);
  if(img->fileSize == 0xFFFFFFFF)
    goto error;

  CloseHandle( targa );
#elif defined (__FreeBSD__)
  struct stat sb;
    
  if(stat(in_file, &sb))
    goto error;
        
  img->fileSize = sb.st_size;
#endif

  if(fseek(fin, img->fileSize - 26, SEEK_SET))
    goto error;
  if(fread( &img->extOffset, 4, 1, fin ) < 1)
    goto error;      
  if(fread( &img->devDirOffset, 4, 1, fin ) < 1)
    goto error;    
  if(fread( &img->signature, 17, 1, fin ) < 1) 
    goto error;
        
  img->signature[16] = '\0';
    
  if(!memcmp(img->signature, "TRUEVISION-XFILE", 15)) 
  {
    if(img->extOffset || img->devDirOffset)
      goto error;
  }
  else 
  {
    img->extOffset = 0;
    img->devDirOffset = 0;
    memset(&img->signature, 0, 17);
  }
    
  /* Start off seeking passed our header for sure. */
  img->imgOffset = 18;
    
  /* If we have an Image ID section we want to seek passed it. */
  if(img->idLength)
    img->imgOffset += img->idLength;
        
  /* If we have a color map section, we want to seek passed it */
  if(img->colorMapType)
    img->imgOffset += (img->colorMapLength * img->cMapEntrySize);
    
  if(fseek(fin, img->imgOffset, SEEK_SET))
    goto error;
    
  if(fread(img->image, img->imgSize, 1, fin) < 1)
    goto error;
    
  fclose(fin);
  return img;

error:
  tga_destroy(&img);
  fclose(fin);
  return NULL;
}


/* 
 * Arguments:
 *  img - Address of a dynamically allocated 'tga_image' structure. Like one 
 *        returned from 'tga_load' tga_image * img; 
 *        img = tga_load("image.tga"); tga_destroy( &img );
 *
 * Purpose:
 *  This is a simple function to clean up a currently loaded TARGA image.
 *
 * Returns:
 *  Nothing.
 */
DLLEXPORT void 
tga_destroy(tga_image **img)
{
  if(!img)
    return ;
        
  if((*img)->image) free((*img)->image);
  free((*img));
  *img = NULL; 
}


libGLi_src/libGLi.c
/*
 * libGLi - A library for loading\saving etc... GLi images
 * Copyright (C) 2005  Frank L. Morales II (altf2o)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * File:
 *  libGLi.c
 *
 * Purpose: 
 *  This is the implementation for our libraries exported functions.
 *
 * File ID:
 *  $Id: libGLi.c 8 2005-03-05 23:43:06Z altf2o $
 */ 
#include <GLi.h>

/*
 * Arguments:
 *  in_file - A valid C-String containing the name of the GLi image to load.
 *
 * Purpose:
 *  This function will dynamically allocate a 'gli_image' structure, filling
 *  it in with the image, whos name is passed in (in_file). If the image is 
 *  compressed this function will uncompress the pixel data. Finally it 
 *  stores the raw pixel data in the 'gli_image' structure' 'image' variable.
 *
 * Returns:
 *   img - A valid 'gli_image' structure, completely filled in.
 *  NULL - On error.
 */
DLLEXPORT gli_image * 
gli_load(const char *in_file)
{
  unsigned long int  uncompSize  = 0;
  gli_image         *img         = NULL;
  GLubyte           *comp_data   = NULL;
  FILE              *fin;
    
  if(!in_file)
    return NULL;
        
  img = malloc(sizeof(*img));
  if(img == NULL)
    return NULL;
  
  memset(img, 0, sizeof(*img));
  
  fin = fopen(in_file, "rb");
  if(fin == NULL) 
  {
    free(img);
    return NULL;
  }
  
  /* First lets snag our header */
  if(fread(img, GLI_HEADER_SIZE, 1, fin) < 1)
    goto error;

  /* Wrong magic byte or version, it's an error */
  if(memcmp(img->mBytes, MAGIC_BYTES, 4) || 
     img->version != GLI_CURRENT_VERSION)
    goto error;
  
  /* The only images we currently support are 8\24\32bits per pixel. */
  if(img->format != GLI_FORMAT_STANDARD &&
     img->format != GLI_FORMAT_32BIT    &&
     img->format != GLI_FORMAT_8BIT)
    goto error;
        
  /* No valid image size, that's an error */
  if(!img->imgSize)
    goto error;
  
  uncompSize = img->imgSize;
  
  /* Allocate the space for our uncompressed raw bitmap data */
  img->image = malloc(img->imgSize);
  if(img->image == NULL)
    goto error;
      
  /* Is this image compressed ? */
  if(img->comp == GLI_IMAGE_COMPRESSED) 
  {    
    /* There better be a valid size for the compressed data */
    if(img->compSize) 
    {
      comp_data = malloc(img->compSize * sizeof(*comp_data));
      if(comp_data != NULL) 
      {
        /* Got our memory, lets read in our compressed image data */
        if(fread(comp_data, img->compSize, 1, fin) == 1)
        {
          if(gli_uncompress(img->image, 
                            &uncompSize, 
                            comp_data, 
                            img->compSize) == GLI_OK) 
          {
            /* Ok no error in uncompressing the data. Did it unpack to the 
             * same size ? If so our raw uncompressed bitmap data is now being
             * stored in img->image, we no longer need "comp_data", free it.
             * If not, there was an error in the uncompression. 
             */
            if(uncompSize == img->imgSize)
              free(comp_data);
            else
              goto error;
          }
          else
            /* Error in gli_uncompress */
            goto error;
        }
        else
          /* Error in fread */
          goto error;
      }
      else
        /* Error mallocing comp_data memory */
        goto error;
    }
    else
      /* No valid value in compSize */
      goto error;
  }
  else 
  {
    /* Uncompressed image */
    if(fread(img->image, img->imgSize, 1, fin) < 1)
      goto error;
  }
  
  fclose(fin);
  return img;
  
error:
  if(comp_data)  free(comp_data);
  if(img->image) free(img->image);
  free(img);
  fclose(fin);
  return NULL;
}

    
/*
 * Arguments:
 *       img - A pointer to a valid, and filled in 'gli_image' structure.
 *  out_file - A valid C-String specifying the file to write to.
 *
 * Purpose:
 *  This function writes our passed in 'gli_image' structure (img) to the
 *  file specified by 'out_file'. This function expects 'img' to be 
 *  completely filled in, compression and all.
 *
 * Returns:
 *     GLI_OK - If everything went ok.
 *  GLI_ERROR - On error.
 */
DLLEXPORT int 
gli_write(const gli_image *img, 
          const char      *out_file)
{
  unsigned int  imgSize;
  FILE         *fout;
    
  if(!img || !out_file || !img->image)
    return GLI_ERROR;
      
  /* Are we compressed ? If so we better have a valid 'compSize' */
  if(img->comp == GLI_IMAGE_COMPRESSED) 
  {
    /* Anything non zero we'll consider as being valid */
    if(img->compSize)
      imgSize = img->compSize;
    else
      return GLI_ERROR;
  }
  else 
  {
    /* Not compressed, we better have valid image size then */
    if(img->imgSize)
      imgSize = img->imgSize;
    else
      return GLI_ERROR;
  }
      
  fout = fopen(out_file, "wb");
  if(fout == NULL)
    return GLI_ERROR;

  if(fwrite(img, GLI_HEADER_SIZE, 1, fout) < 1) 
  {
    fclose(fout);
    return GLI_ERROR;
  }
  
  if(fwrite(img->image, imgSize, 1, fout) < 1) 
  {
    fclose(fout);
    return GLI_ERROR;
  }
  
  fclose(fout);
  return GLI_OK;
}


/* 
 * Arguments:
 *  img - This should be the 'address' of a valid dynamically allocated
 *        'gli_image' structure. I.e.: 
 *        gli_image * img; img = gli_load("image.gli"); gli_destroy( &img );
 *                         
 * Purpose:
 *  This function destroys a dynamically allocated 'gli_image' structure,
 *  like the one returned from 'gli_load'.
 *
 * Returns:
 *   Nothing.
 */
DLLEXPORT void 
gli_destroy(gli_image **img)
{
  if(!img)
    return ;
      
  free((*img)->image);
  free((*img));
  *img = NULL;
}


/*
 * Arguments:
 *  dest_buff - A pointer to a destination buffer large enough to hold the
 *              compressed data.
 *  dest_len  - The length of our destination buffer.
 *  src_buff  - A pointer to a source buffer that contains our raw pixel data.
 *  src_len   - The length of our source buffer.
 *
 * Purpose:
 *  This function is simply a wrapper to the zlib function call, "compress2",
 *  which uses a default compression of Z_BEST_COMPRESSION. If compress2
 *  works ok we map it's return value to our own GLI_OK, otherwise we map 
 *  it to GLI_ERROR.
 *
 * Returns:
 *     GLI_OK - If everything went ok.
 *  GLI_ERROR - On error.
 */
DLLEXPORT int
gli_compress(char              *dest_buff, 
             unsigned long int *dest_len,
             const char        *src_buff, 
             unsigned long int  src_len)
{
  return (compress2(dest_buff, 
                    dest_len, 
                    src_buff, 
                    src_len, 
                    Z_BEST_COMPRESSION) == Z_OK) ? GLI_OK : GLI_ERROR;
}
                              

/* 
 * Arguments:
 *  dest_buff - A pointer to a destination buffer large enough to hold our
 *              uncompressed data.
 *  dest_len  - The length of our destination buffer.
 *  src_buff  - A pointer to our source buffer that contains our compressed
 *              pixel data.
 *  src_len   - The length of our source buffer.
 *
 * Purpose:
 *  This function is simply a wrapper to the zlib function call, "uncompress"
 *  which uncompresses our source buffer, and stores it in our destination
 *  buffer. Again we map the return of uncompress to a value of our own.
 *
 * Returns:
 *     GLI_OK - If everything went ok.
 *  GLI_ERROR - On error.
 */
DLLEXPORT int  
gli_uncompress(char              *dest_buff, 
               unsigned long int *dest_len,
               const char        *src_buff, 
               unsigned long int  src_len )
{
  return (uncompress(dest_buff, 
                     dest_len, 
                     src_buff, 
                     src_len) == Z_OK) ? GLI_OK : GLI_ERROR;
}

                                                                
/*
 * Arguments:
 *  out_file - This should be a valid C-String containing the name of the 
 *             file to write to.
 *  author   - This should be a valid C-string (23 char max) containing 
 *             the author's name.
 *  stamp    - This should be a valid C-string (27 char max) containing a
 *             data\time stamp.
 *  compress - This should be GLI_IMAGE_COMPRESS if you want to use 
 *             compression. Otherwise you should use: GLI_IMAGE_DONTCOMPRESS.
 *             All values other than: GLI_IMAGE_COMPRESS are assumed to mean
 *             do NOT use compression.
 *
 * Purpose:
 *  This function is specifically for writing what's currently in your OpenGL
 *  windows front buffer (a screenshot if you will). It will determin the
 *  current bits per pixel of your current window, and fill in a valid 
 *  'gli_image' structure. If compression is wanted it does all that as well.
 *  Then it simply passes the filled in 'gli_image' structure, plus the file
 *  you want to write to (out_file) to 'gli_write' which writes the actual
 *  image to disk.
 *
 * Returns:
 *     GLI_OK - If everything went ok.
 *  GLI_ERROR - On error.
 */
DLLEXPORT int 
gli_writeOGLBuffer(const char    *out_file, 
                   const char    *author, 
                   const char    *stamp, 
                   unsigned char  compress)
{
  unsigned long int  comp_size;
  gli_image          img;
  GLubyte           *comp_data;
  GLenum             lastBuff;
  GLint              viewport[4];
  GLint              rBits;
  GLint              gBits;
  GLint              bBits;
  GLint              aBits;

  memset(&img, 0, sizeof(img));
  
  /* Compute how many bits per pixel we're working with */
  glGetIntegerv(GL_RED_BITS, &rBits);
  glGetIntegerv(GL_GREEN_BITS, &gBits);
  glGetIntegerv(GL_BLUE_BITS, &bBits);
  glGetIntegerv(GL_ALPHA_BITS, &aBits);
          
  img.pxlDepth = (rBits + gBits + bBits + aBits) / 8;
                
  /* Right now we're only supporting 8\24\32bits per pixel. */
  if(img.pxlDepth != 3 && img.pxlDepth != 4 && img.pxlDepth != 1)
    return GLI_ERROR;
                
  /* Assign our images format */
  img.format = (img.pxlDepth == 3) ? GLI_FORMAT_STANDARD : 
               (img.pxlDepth == 4) ? GLI_FORMAT_32BIT    : 
               GLI_FORMAT_8BIT;
                
  /* Now we see how big our current window is. viewport[2] contains the 
   * width while viewport[3] contains the height. 
   */        
  glGetIntegerv(GL_VIEWPORT, viewport);
  
  img.imgSize = viewport[2] * viewport[3] * img.pxlDepth;
  
  img.image = malloc(img.imgSize * sizeof(*img.image));
  if(img.image == NULL) 
    return GLI_ERROR;
  
  /* Here we tell OpenGL to keep things tightly packed when we read from the 
   * color buffer and store to memory.
   */
  glPixelStorei(GL_PACK_ALIGNMENT, 1);
  glPixelStorei(GL_PACK_ROW_LENGTH, 0);
  glPixelStorei(GL_PACK_SKIP_ROWS, 0);
  glPixelStorei(GL_PACK_SKIP_PIXELS, 0);

  /* We want to preserve the current buffer, just incase the current one
   * isn't GL_FRONT 
   */
  glGetIntegerv(GL_READ_BUFFER, &lastBuff);
  glReadBuffer(GL_FRONT);
  glReadPixels(0, 
               0, 
               viewport[2], 
               viewport[3], 
               img.format, 
               GL_UNSIGNED_BYTE, 
               img.image);
  glReadBuffer(lastBuff);
  
  /* Here we setup the rest of our structures header */
  memcpy(img.mBytes, MAGIC_BYTES, 4);
  img.version = GLI_CURRENT_VERSION;
  img.width = viewport[2];
  img.height = viewport[3];
  strncpy(img.author, author, 23);
  strncpy(img.stamp, stamp, 27);
  
  if(compress == GLI_IMAGE_COMPRESS) 
  {
    /* Lets get a second buffer for compressing the data */
    comp_data = malloc(img.imgSize * sizeof(*comp_data));
    if(comp_data == NULL) 
    {
      free(img.image);
      return GLI_ERROR;
    }
    else 
    {
      comp_size = img.imgSize;
      if(gli_compress(comp_data, &comp_size, img.image, img.imgSize) == GLI_OK)
      {
        /* No longer need our raw bitmap data, free it */
        free(img.image);

        /* Set our image data to point to our compressed data buffer */
        img.image = comp_data;
        comp_data = NULL;
        
        img.comp = GLI_IMAGE_COMPRESSED;
        img.compSize = comp_size;
      }
      else 
      {
        /* Error compressing. Simply store it as an uncompressed image */
        img.comp = GLI_IMAGE_NOTCOMPRESSED;
        img.compSize = 0;
        free(comp_data);
      }
    }    
  }
  else 
  { 
    /* Passed in argument "compress" was NOT: GLI_IMAGE_COMPRESS */  
    img.comp = GLI_IMAGE_NOTCOMPRESSED;
    img.compSize = 0;
  }
  
  /* Now we write our actual GLi image. */
  if(gli_write(&img, out_file) != GLI_OK) 
  {
    free(img.image);
    return GLI_ERROR;
  }
  
  /* Everything went well. The only thing that's dynamically allocated
   * is this. 
   */
  free(img.image);
  return GLI_OK;
}
